### PR TITLE
Make Mongoose raise an error when no odbc pools configured

### PR DESCRIPTION
**Pull request by kanes115**
What I am not sure about, @fenek:
* The way I stop Mongoose - including extra file in ejabberd_app.erl and calling exit with a 
?STATUS_ERROR (maybe should be ?STATUS_USAGE ?)
* I decided to stop executing program in ejabberd_app.erl, because it is the highest level it can be done on. What worries me is - it is not done this way now

### CODE REVIEW

File: apps/ejabberd/src/ejabberd_rdbms.erl [hunk 1]
```diff
@@ -30,7 +30,7 @@
 -export([start/0, start_pool/1, stop_pool/1, pools/0]).
 -include("ejabberd.hrl").
 
--spec start() -> 'ok' | {'error', 'lager_not_running'}.
+-spec start() -> 'ok' | {'error', Reason :: atom()}.
```
Comment (ludwikbukowski) [inline apps/ejabberd/src/ejabberd_rdbms.erl]: what do you think about this:
```
-spec start() -> 'ok' | {'error', 'lager_not_running'} | {'error', `no_pools_configured`}.
```
there are only two types of error so maybe we should leave it more informative? :)

File: apps/ejabberd/src/ejabberd_app.erl [hunk 2]
```diff
@@ -75,6 +75,18 @@ start(normal, _Args) ->
 start(_, _) ->
     {error, badarg}.
 
+
+
+start_rdbms() ->
+    case ejabberd_rdbms:start() of
+        {error, no_pools_configured} ->
+            ?ERROR_MSG("RDBMS connection has been configured with no pools.", []),
+            exit(?STATUS_ERROR);
```
Comment (kanes115) [inline apps/ejabberd/src/ejabberd_app.erl]: Should these exits be replaced with `init:stop()`, @fenek ?

File: apps/ejabberd/src/ejabberd_rdbms.erl [hunk 3]
```diff
@@ -82,3 +83,28 @@ normalize_type(mssql) ->
...
 normalize_type(_) ->
     "generic".
+
+% @doc Function checks whether user configured odbc connection.
+% It relies on ejabberd_config that adds option `odbc_configured`
+% with a value of true when odbc connection was defined.
+is_connection_defined() ->
+    case ejabberd_config:get_local_option(odbc_configured) of
+        true -> true;
+	_ -> false
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_rdbms.erl]: According to **Elvis**:
> Line 93 has a tab at column 0.

File: apps/ejabberd/src/ejabberd_config.erl [hunk 4]
```diff
@@ -516,7 +516,7 @@ process_term(Term, State) ->
                             process_host_term(T, list_to_binary(Host), S) end,
                         State, Terms);
         {pool, odbc, _PoolName} ->
-            State;
+	    State;
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_config.erl]: According to **Elvis**:
> Line 519 has a tab at column 0.

File: apps/ejabberd/src/ejabberd_rdbms.erl [hunk 5]
```diff
@@ -82,3 +83,28 @@ normalize_type(mssql) ->
...
+    end.
+
+% @doc Starts pools when odbc connection was configured
+maybe_start_pools(Pools) ->
+    case is_connection_defined() of
+	false ->
+	    ok;
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_rdbms.erl]: According to **Elvis**:
> Line 100 has a tab at column 0.

File: apps/ejabberd/src/ejabberd_rdbms.erl [hunk 6]
```diff
@@ -82,3 +83,28 @@ normalize_type(mssql) ->
...
+	true ->
+	    start_pools(Pools)
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_rdbms.erl]: According to **Elvis**:
> Line 102 has a tab at column 0.

File: apps/ejabberd/src/ejabberd_config.erl [hunk 7]
```diff
@@ -597,6 +597,7 @@ process_term_for_hosts_and_pools(Term = {Key, _Val}, State) ->
     BKey = atom_to_binary(Key, utf8),
     case get_key_group(BKey, Key) of
         odbc ->
+	    add_local_option(odbc_configured, true),
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_config.erl]: According to **Elvis**:
> Line 600 has a tab at column 0.

File: apps/ejabberd/src/ejabberd_rdbms.erl [hunk 8]
```diff
@@ -82,3 +83,28 @@ normalize_type(mssql) ->
...
+        true -> true;
+	_ -> false
+    end.
+
+% @doc Starts pools when odbc connection was configured
+maybe_start_pools(Pools) ->
+    case is_connection_defined() of
+	false ->
+	    ok;
+	true ->
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_rdbms.erl]: According to **Elvis**:
> Line 101 has a tab at column 0.

File: apps/ejabberd/src/ejabberd_rdbms.erl [hunk 9]
```diff
@@ -82,3 +83,28 @@ normalize_type(mssql) ->
...
+is_connection_defined() ->
+    case ejabberd_config:get_local_option(odbc_configured) of
+        true -> true;
+	_ -> false
+    end.
+
+% @doc Starts pools when odbc connection was configured
+maybe_start_pools(Pools) ->
+    case is_connection_defined() of
+	false ->
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_rdbms.erl]: According to **Elvis**:
> Line 99 has a tab at column 0.

File: apps/ejabberd/src/ejabberd_config.erl [hunk 10]
```diff
@@ -597,13 +597,22 @@ process_term_for_hosts_and_pools(Term = {Key, _Val}, State) ->
...
                         State, State#state.hosts)
     end.
 
+check_pools([]) ->
+    ?CRITICAL_MSG("Config file invalid: ODBC defined with no pools", []),
+    exit(no_odbc_pools);
+check_pools(Pools) when is_list(Pools) ->
+    ok;
+check_pools(_) ->
+    {error, bad_arg}.
```
Comment (fenek) [inline apps/ejabberd/src/ejabberd_config.erl]: The convention is to simply use atom `badarg` without underscore. Anyway, this clause is not needed, as this function will fail anyway with a non-list argument and will include the invalid argument value in error description for debugging.

**Comment by kanes115** [comment]
@fenek, I think it's ready now 👍

**Comment by kanes115** [comment]
@fenek, I think it's ready now 👍

**Comment by fenek** [comment]
We've decided to add regression tests for this.

**Comment by kanes115** [comment]
@fenek, tests implemented

**Comment by fenek** [comment]
You forgot to commit actual config files. :)

**Pull request merged**